File: simpletranscribe-win/SimpleTranscribe/Models/SupportedLanguages.cs
namespace SimpleTranscribe.Models;

public class SupportedLanguage
{
    public string Code { get; init; } = "";
    public string DisplayName { get; init; } = "";
    public override string ToString() => DisplayName;
}

public static class SupportedLanguages
{
    public static readonly List<SupportedLanguage> Whisper = new()
    {
        new() { Code = "auto", DisplayName = "Auto Detect" },
        new() { Code = "en",   DisplayName = "English" },
        new() { Code = "zh",   DisplayName = "Chinese" },
        new() { Code = "de",   DisplayName = "German" },
        new() { Code = "es",   DisplayName = "Spanish" },
        new() { Code = "ru",   DisplayName = "Russian" },
        new() { Code = "ko",   DisplayName = "Korean" },
        new() { Code = "fr",   DisplayName = "French" },
        new() { Code = "ja",   DisplayName = "Japanese" },
        new() { Code = "pt",   DisplayName = "Portuguese" },
        new() { Code = "tr",   DisplayName = "Turkish" },
        new() { Code = "pl",   DisplayName = "Polish" },
        new() { Code = "nl",   DisplayName = "Dutch" },
        new() { Code = "ar",   DisplayName = "Arabic" },
        new() { Code = "it",   DisplayName = "Italian" },
        new() { Code = "sv",   DisplayName = "Swedish" },
        new() { Code = "hi",   DisplayName = "Hindi" },
        new() { Code = "da",   DisplayName = "Danish" },
        new() { Code = "fi",   DisplayName = "Finnish" },
        new() { Code = "he",   DisplayName = "Hebrew" },
        new() { Code = "uk",   DisplayName = "Ukrainian" },
        new() { Code = "cs",   DisplayName = "Czech" },
        new() { Code = "el",   DisplayName = "Greek" },
        new() { Code = "hr",   DisplayName = "Croatian" },
        new() { Code = "hu",   DisplayName = "Hungarian" },
        new() { Code = "ro",   DisplayName = "Romanian" },
        new() { Code = "sk",   DisplayName = "Slovak" },
        new() { Code = "no",   DisplayName = "Norwegian" },
        new() { Code = "bg",   DisplayName = "Bulgarian" },
        new() { Code = "id",   DisplayName = "Indonesian" },
        new() { Code = "ms",   DisplayName = "Malay" },
        new() { Code = "th",   DisplayName = "Thai" },
        new() { Code = "vi",   DisplayName = "Vietnamese" },
        new() { Code = "ca",   DisplayName = "Catalan" },
        new() { Code = "gl",   DisplayName = "Galician" },
        new() { Code = "sl",   DisplayName = "Slovenian" },
        new() { Code = "et",   DisplayName = "Estonian" },
        new() { Code = "lv",   DisplayName = "Latvian" },
        new() { Code = "lt",   DisplayName = "Lithuanian" },
        new() { Code = "af",   DisplayName = "Afrikaans" },
        new() { Code = "az",   DisplayName = "Azerbaijani" },
        new() { Code = "be",   DisplayName = "Belarusian" },
        new() { Code = "bn",   DisplayName = "Bengali" },
        new() { Code = "bs",   DisplayName = "Bosnian" },
        new() { Code = "cy",   DisplayName = "Welsh" },
        new() { Code = "eu",   DisplayName = "Basque" },
        new() { Code = "fa",   DisplayName = "Persian" },
        new() { Code = "gu",   DisplayName = "Gujarati" },
        new() { Code = "hy",   DisplayName = "Armenian" },
        new() { Code = "is",   DisplayName = "Icelandic" },
        new() { Code = "ka",   DisplayName = "Georgian" },
        new() { Code = "kk",   DisplayName = "Kazakh" },
        new() { Code = "km",   DisplayName = "Khmer" },
        new() { Code = "kn",   DisplayName = "Kannada" },
        new() { Code = "lo",   DisplayName = "Lao" },
        new() { Code = "lb",   DisplayName = "Luxembourgish" },
        new() { Code = "mk",   DisplayName = "Macedonian" },
        new() { Code = "ml",   DisplayName = "Malayalam" },
        new() { Code = "mn",   DisplayName = "Mongolian" },
        new() { Code = "mr",   DisplayName = "Marathi" },
        new() { Code = "mt",   DisplayName = "Maltese" },
        new() { Code = "my",   DisplayName = "Myanmar" },
        new() { Code = "ne",   DisplayName = "Nepali" },
        new() { Code = "pa",   DisplayName = "Punjabi" },
        new() { Code = "si",   DisplayName = "Sinhala" },
        new() { Code = "sq",   DisplayName = "Albanian" },
        new() { Code = "sr",   DisplayName = "Serbian" },
        new() { Code = "su",   DisplayName = "Sundanese" },
        new() { Code = "sw",   DisplayName = "Swahili" },
        new() { Code = "ta",   DisplayName = "Tamil" },
        new() { Code = "te",   DisplayName = "Telugu" },
        new() { Code = "tg",   DisplayName = "Tajik" },
        new() { Code = "tl",   DisplayName = "Tagalog" },
        new() { Code = "tt",   DisplayName = "Tatar" },
        new() { Code = "ur",   DisplayName = "Urdu" },
        new() { Code = "uz",   DisplayName = "Uzbek" },
        new() { Code = "yi",   DisplayName = "Yiddish" },
        new() { Code = "yo",   DisplayName = "Yoruba" },
        new() { Code = "yue",  DisplayName = "Cantonese" },
        new() { Code = "jw",   DisplayName = "Javanese" },
        new() { Code = "haw",  DisplayName = "Hawaiian" },
        new() { Code = "mi",   DisplayName = "Maori" },
        new() { Code = "ht",   DisplayName = "Haitian Creole" },
        new() { Code = "ha",   DisplayName = "Hausa" },
        new() { Code = "sn",   DisplayName = "Shona" },
        new() { Code = "sd",   DisplayName = "Sindhi" },
        new() { Code = "sa",   DisplayName = "Sanskrit" },
        new() { Code = "la",   DisplayName = "Latin" },
    };

    public static readonly HashSet<string> ParakeetV3 = new()
    {
        "en", "de", "fr", "es", "it", "pt", "nl", "pl", "cs", "sk",
        "hr", "sl", "bg", "ro", "hu", "da", "sv", "fi", "no", "et",
        "lv", "lt", "uk", "ru", "el", "ca", "gl"
    };

    public static readonly HashSet<string> ParakeetV2 = new() { "en" };

    public static HashSet<string>? SupportedCodes(string modelId)
    {
        if (modelId.Contains("parakeet-tdt-0.6b-v2")) return ParakeetV2;
        if (modelId.Contains("parakeet-tdt-0.6b-v3")) return ParakeetV3;
        return null;
    }

    public static List<SupportedLanguage> Available(string modelId)
    {
        var allowed = SupportedCodes(modelId);
        if (allowed == null) return Whisper;
        return Whisper.Where(l => l.Code == "auto" || allowed.Contains(l.Code)).ToList();
    }
}

File: simpletranscribe-mac/simpletranscribe/Models/SupportedLanguages.swift
import Foundation

struct SupportedLanguage: Identifiable {
    let code: String
    let displayName: String
    var id: String { code }
}

enum SupportedLanguages {
    // Full Whisper language list (~100 languages from whisper.cpp's language table)
    static let whisper: [SupportedLanguage] = [
        SupportedLanguage(code: "auto", displayName: "Auto Detect"),
        SupportedLanguage(code: "en",   displayName: "English"),
        SupportedLanguage(code: "zh",   displayName: "Chinese"),
        SupportedLanguage(code: "de",   displayName: "German"),
        SupportedLanguage(code: "es",   displayName: "Spanish"),
        SupportedLanguage(code: "ru",   displayName: "Russian"),
        SupportedLanguage(code: "ko",   displayName: "Korean"),
        SupportedLanguage(code: "fr",   displayName: "French"),
        SupportedLanguage(code: "ja",   displayName: "Japanese"),
        SupportedLanguage(code: "pt",   displayName: "Portuguese"),
        SupportedLanguage(code: "tr",   displayName: "Turkish"),
        SupportedLanguage(code: "pl",   displayName: "Polish"),
        SupportedLanguage(code: "nl",   displayName: "Dutch"),
        SupportedLanguage(code: "ar",   displayName: "Arabic"),
        SupportedLanguage(code: "it",   displayName: "Italian"),
        SupportedLanguage(code: "sv",   displayName: "Swedish"),
        SupportedLanguage(code: "hi",   displayName: "Hindi"),
        SupportedLanguage(code: "da",   displayName: "Danish"),
        SupportedLanguage(code: "fi",   displayName: "Finnish"),
        SupportedLanguage(code: "he",   displayName: "Hebrew"),
        SupportedLanguage(code: "uk",   displayName: "Ukrainian"),
        SupportedLanguage(code: "cs",   displayName: "Czech"),
        SupportedLanguage(code: "el",   displayName: "Greek"),
        SupportedLanguage(code: "hr",   displayName: "Croatian"),
        SupportedLanguage(code: "hu",   displayName: "Hungarian"),
        SupportedLanguage(code: "ro",   displayName: "Romanian"),
        SupportedLanguage(code: "sk",   displayName: "Slovak"),
        SupportedLanguage(code: "no",   displayName: "Norwegian"),
        SupportedLanguage(code: "bg",   displayName: "Bulgarian"),
        SupportedLanguage(code: "id",   displayName: "Indonesian"),
        SupportedLanguage(code: "ms",   displayName: "Malay"),
        SupportedLanguage(code: "th",   displayName: "Thai"),
        SupportedLanguage(code: "vi",   displayName: "Vietnamese"),
        SupportedLanguage(code: "ca",   displayName: "Catalan"),
        SupportedLanguage(code: "gl",   displayName: "Galician"),
        SupportedLanguage(code: "sl",   displayName: "Slovenian"),
        SupportedLanguage(code: "et",   displayName: "Estonian"),
        SupportedLanguage(code: "lv",   displayName: "Latvian"),
        SupportedLanguage(code: "lt",   displayName: "Lithuanian"),
        SupportedLanguage(code: "af",   displayName: "Afrikaans"),
        SupportedLanguage(code: "az",   displayName: "Azerbaijani"),
        SupportedLanguage(code: "be",   displayName: "Belarusian"),
        SupportedLanguage(code: "bn",   displayName: "Bengali"),
        SupportedLanguage(code: "bs",   displayName: "Bosnian"),
        SupportedLanguage(code: "cy",   displayName: "Welsh"),
        SupportedLanguage(code: "eu",   displayName: "Basque"),
        SupportedLanguage(code: "fa",   displayName: "Persian"),
        SupportedLanguage(code: "gl",   displayName: "Galician"),
        SupportedLanguage(code: "gu",   displayName: "Gujarati"),
        SupportedLanguage(code: "hy",   displayName: "Armenian"),
        SupportedLanguage(code: "is",   displayName: "Icelandic"),
        SupportedLanguage(code: "ka",   displayName: "Georgian"),
        SupportedLanguage(code: "kk",   displayName: "Kazakh"),
        SupportedLanguage(code: "km",   displayName: "Khmer"),
        SupportedLanguage(code: "kn",   displayName: "Kannada"),
        SupportedLanguage(code: "lo",   displayName: "Lao"),
        SupportedLanguage(code: "lb",   displayName: "Luxembourgish"),
        SupportedLanguage(code: "mk",   displayName: "Macedonian"),
        SupportedLanguage(code: "ml",   displayName: "Malayalam"),
        SupportedLanguage(code: "mn",   displayName: "Mongolian"),
        SupportedLanguage(code: "mr",   displayName: "Marathi"),
        SupportedLanguage(code: "mt",   displayName: "Maltese"),
        SupportedLanguage(code: "my",   displayName: "Myanmar"),
        SupportedLanguage(code: "ne",   displayName: "Nepali"),
        SupportedLanguage(code: "nn",   displayName: "Nynorsk"),
        SupportedLanguage(code: "oc",   displayName: "Occitan"),
        SupportedLanguage(code: "pa",   displayName: "Punjabi"),
        SupportedLanguage(code: "ps",   displayName: "Pashto"),
        SupportedLanguage(code: "si",   displayName: "Sinhala"),
        SupportedLanguage(code: "sq",   displayName: "Albanian"),
        SupportedLanguage(code: "sr",   displayName: "Serbian"),
        SupportedLanguage(code: "su",   displayName: "Sundanese"),
        SupportedLanguage(code: "sw",   displayName: "Swahili"),
        SupportedLanguage(code: "ta",   displayName: "Tamil"),
        SupportedLanguage(code: "te",   displayName: "Telugu"),
        SupportedLanguage(code: "tg",   displayName: "Tajik"),
        SupportedLanguage(code: "tk",   displayName: "Turkmen"),
        SupportedLanguage(code: "tl",   displayName: "Tagalog"),
        SupportedLanguage(code: "tt",   displayName: "Tatar"),
        SupportedLanguage(code: "ur",   displayName: "Urdu"),
        SupportedLanguage(code: "uz",   displayName: "Uzbek"),
        SupportedLanguage(code: "yi",   displayName: "Yiddish"),
        SupportedLanguage(code: "yo",   displayName: "Yoruba"),
        SupportedLanguage(code: "yue",  displayName: "Cantonese"),
        SupportedLanguage(code: "jw",   displayName: "Javanese"),
        SupportedLanguage(code: "haw",  displayName: "Hawaiian"),
        SupportedLanguage(code: "sa",   displayName: "Sanskrit"),
        SupportedLanguage(code: "bo",   displayName: "Tibetan"),
        SupportedLanguage(code: "br",   displayName: "Breton"),
        SupportedLanguage(code: "fo",   displayName: "Faroese"),
        SupportedLanguage(code: "la",   displayName: "Latin"),
        SupportedLanguage(code: "ln",   displayName: "Lingala"),
        SupportedLanguage(code: "mg",   displayName: "Malagasy"),
        SupportedLanguage(code: "mi",   displayName: "Maori"),
        SupportedLanguage(code: "ht",   displayName: "Haitian Creole"),
        SupportedLanguage(code: "ha",   displayName: "Hausa"),
        SupportedLanguage(code: "sn",   displayName: "Shona"),
        SupportedLanguage(code: "sd",   displayName: "Sindhi"),
    ]

    /// Languages supported by Parakeet V3 (27 EU + RU/UK)
    static let parakeetV3: Set<String> = [
        "en", "de", "fr", "es", "it", "pt", "nl", "pl", "cs", "sk",
        "hr", "sl", "bg", "ro", "hu", "da", "sv", "fi", "no", "et",
        "lv", "lt", "uk", "ru", "el", "ca", "gl"
    ]

    /// Languages supported by Parakeet V2 (English only)
    static let parakeetV2: Set<String> = ["en"]

    /// Returns the set of allowed language codes for a given model ID.
    /// Returns nil if the model supports all Whisper languages.
    static func supportedCodes(for modelID: String) -> Set<String>? {
        if modelID.contains("parakeet-tdt-0.6b-v2") { return parakeetV2 }
        if modelID.contains("parakeet-tdt-0.6b-v3") { return parakeetV3 }
        return nil
    }

    /// Filtered language list for a given model ID.
    static func available(for modelID: String) -> [SupportedLanguage] {
        guard let allowed = supportedCodes(for: modelID) else { return whisper }
        return whisper.filter { $0.code == "auto" || allowed.contains($0.code) }
    }
}
